File: solutions/3981-jump-game-ix/jump-game-ix.cs
public class Solution {
    public int[] MaxValue(int[] nums) {
        int n = nums.Length;
        int[] ans = new int[n];

        List<(int Value, int Left, int Right)> stack = new();

        for (int i = 0; i < n; i++) {
            var curr = (Value: nums[i], Left: i, Right: i);

            while (stack.Count > 0 && stack[^1].Value > nums[i]) {
                var top = stack[^1];
                stack.RemoveAt(stack.Count - 1);
                curr = (Math.Max(curr.Value, top.Value), top.Left, curr.Right);
            }

            stack.Add(curr);
        }

        for (int i = 0; i < stack.Count; i++) {
            for (int j = stack[i].Left; j <= stack[i].Right; j++) {
                ans[j] = stack[i].Value;
            }
        }

        return ans;
    }
}

File: solutions/3981-jump-game-ix/jump-game-ix.js
var maxValue = function (nums) {
    const n = nums.length;
    const ans = new Array(n);

    const stack = [];

    for (let i = 0; i < n; i++) {
        let curr = {
            value: nums[i],
            left: i,
            right: i,
        };

        while (stack.length > 0 && stack.at(-1).value > nums[i]) {
            const top = stack.pop();
            curr = {
                value: Math.max(curr.value, top.value),
                left: top.left,
                right: curr.right,
            };
        }

        stack.push(curr);
    }

    for (let i = 0; i < stack.length; i++) {
        for (let j = stack[i].left; j <= stack[i].right; j++) {
            ans[j] = stack[i].value;
        }
    }

    return ans;
};